firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+from firedrake.petsc import PETSc
+from firedrake.functionspace import MixedFunctionSpace
+from ufl.duals import is_primal, is_dual
+from pyop2.mpi import internal_comm
+from functools import cached_property
+from itertools import chain
+
+
+def _is_primal_or_dual(local_spaces, ensemble):
+    """
+    We need to test primal or dual collectively over all ensemble
+    ranks otherwise some may fail when others pass.
+
+    Parameters
+    ----------
+
+    local_spaces : Collection
+        The list of :function:`firedrake.FunctionSpace` on the local ensemble.comm.
+    ensemble : :class:`firedrake.Ensemble`
+        The communicator to test collectively over.
+
+    Returns
+    -------
+    str : Description of the collective space type:
+        'primal' if all spaces on all ranks are primal.
+        'dual' if all spaces on all ranks are dual.
+        'invalid' if any "spaces" are not primal or dual, or if some are primal or some are dual.
+
+    Raises
+    ------
+    ValueError : If local_spaces are not all defined on ensemble.comm
+    """
+    local_comm_valid = all(
+        V.mesh().comm is ensemble.comm for V in local_spaces)
+    global_comm_valid = all(
+        ensemble.ensemble_comm.allgather(local_comm_valid))
+    if not global_comm_valid:
+        raise ValueError(
+            "All spaces must be defined on the ensemble.comm")
+
+    local_types = []
```

Comment:
Why not have:
```py
local_types = set()
for V in local_spaces:
        if is_primal(V):
            local_types.add('primal')
        elif is_dual(V):
            local_types.add('dual')
        else:
            # can this ever happen?
            local_types.add('invalid')

if len(local_types) > 1:
    local_type = "invalid"
else:
    local_type, = local_types
```
Then you only have to broadcast one string and can avoid needing `chain()`.

Reply:
Much nicer, thanks.